kubetest/gke.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package main / gke.go provides the Google Container Engine (GKE)
+// kubetest deployer via newGKE().
+//
+// TODO(zmerlynn): Pull this out to a separate package?
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+const (
+	defaultPool = "default"
+	e2eAllow    = "tcp:22,tcp:80,tcp:8080,tcp:30000-32767,udp:30000-32767"
+)
+
+var (
+	gkeCluster         = flag.String("gke-cluster", "", "(gke only) Cluster name. Must be set for GKE.")
+	gkeAdditionalZones = flag.String("gke-additional-zones", "", "(gke only) List of additional Google Compute Engine zones to use. Clusters are created symmetrically across zones by default, see --gke-shape for details.")
+	gkeShape           = flag.String("gke-shape", `{"default":{"Nodes":3,"MachineType":"n1-standard-2"}}`, `(gke only) A JSON description of node pools to create. The node pool 'default' is required and used for initial cluster creation. All node pools are symmetric across zones, so the cluster total node count is {total nodes in --gke-shape} * {1 + (length of --gke-additional-zones)}. Example: '{"default":{"Nodes":999,"MachineType:":"n1-standard-1"},"heapster":{"Nodes":1, "MachineType":"n1-standard-8"}}`)
```

Comment:
I think this one is a little more GKE specific, and scalability is the only one I anticipated using it with (because there's a funny back to launch a node pool for Heapster). There's not an easy way to do this in OSS right now, either. Let's keep it GKE specific and come back?